packages/mongodb-chatbot-eval/src/generate/generateConversationData.ts

```diff
@@ -0,0 +1,150 @@
+import { ConversationGeneratedData } from "./GeneratedDataStore";
+import { ConversationsService, ObjectId } from "mongodb-chatbot-server";
+import { logger } from "mongodb-rag-core";
+import { GenerateDataFunc } from "./GenerateDataFunc";
+import { ConversationTestCase, SomeTestCase } from "./TestCase";
+import { strict as assert } from "assert";
+import axios from "axios";
+
+export interface MakeGenerateConversationDataParams {
+  /**
+    Same `ConversationsService` instance used in the chatbot.
+    The function uses the service to create conversations and add initial messages.
+   */
+  conversations: ConversationsService;
+
+  /**
+    URL for the server you're evaluating.
+    @default "http://localhost:3000/api/v1/"
+  */
+  apiBaseUrl?: string;
+
+  /**
+    HTTP headers to include in the request to add a message to a conversation.
+    @example
+    ```json
+    {
+      "Authorization": "Bearer <some token>",
+    }
+    ```
+  */
+  httpHeaders?: Record<string, string>;
+
+  /**
+    Number of milliseconds to sleep between each conversation.
+    This can be useful if there's a rate limit on some components of the chatbot.
+    Often LLM APIs have a rate limit, for example.
+    @default 0
+   */
+  sleepMs?: number;
+}
+/**
+  Generate conversation data from test cases.
+ */
+export const makeGenerateConversationData = function ({
+  conversations,
+  httpHeaders = {},
+  apiBaseUrl = "http://localhost:3000/api/v1/",
+  sleepMs = 0,
+}: MakeGenerateConversationDataParams): GenerateDataFunc {
+  return async function ({
+    testCases,
+    runId,
+  }: {
+    testCases: SomeTestCase[];
+    runId: ObjectId;
+  }): Promise<{
+    generatedData: ConversationGeneratedData[];
+    failedCases: ConversationTestCase[];
+  }> {
+    // FIXME: how to go away from this to something more elegant/typescripty?
+    const convoTestCases = testCases as ConversationTestCase[];
```

Comment:
perf, was looking for something just like that. thx!

Reply:
still needed to add the `as ConversationTestCase[]` at the end but at least now we have a decent guarantee that this is the case:

```ts
const convoTestCases = testCases.filter((testCase) =>
      isConversationTestCase(testCase)
    ) as ConversationTestCase[];
```

Reply:
My initial testing for this wasn't totally right since there's some trickiness with type inference across scopes. Basically the type guard needs to be on the filter function itself rather than inferred from the function body.

There are two way to avoid the type assertion here:

1. The "clean" way is to use the type guard directly as the filter function. It's probably fine but might break if we ever added more args to `isConversationTestCase` (which IMO is unlikely given the nature of a type guard function)
```ts
const convoTestCases = testCases.filter(isConversationTestCase);
```

2. The "explicit" way re-declares the type guard on the filter function. It's a bit repetitive with the type guard function definition but avoids the type assertion
```ts
const convoTestCases = testCases.filter(
  (testCase): testCase is ConversationTestCase =>
    isConversationTestCase(testCase)
);
```